File: back_velora/src/main/java/com/velora/inventory/InventoryController.java
package com.velora.inventory;

import java.util.List;
import java.util.UUID;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;

import com.velora.inventory.dto.InventoryMovementRequest;
import com.velora.inventory.dto.InventoryMovementResponse;
import com.velora.inventory.dto.InventoryStockResponse;
import com.velora.inventory.dto.WarehouseRequest;
import com.velora.inventory.dto.WarehouseResponse;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/inventory")
public class InventoryController {

    private final InventoryService inventoryService;

    public InventoryController(InventoryService inventoryService) {
        this.inventoryService = inventoryService;
    }

    @GetMapping("/warehouses")
    public List<WarehouseResponse> warehouses(
            @AuthenticationPrincipal Jwt jwt
    ) {
        return inventoryService.listWarehouses(actorId(jwt));
    }

    @PostMapping("/warehouses")
    public WarehouseResponse createWarehouse(
            @Valid @RequestBody WarehouseRequest request,
            @AuthenticationPrincipal Jwt jwt
    ) {
        return inventoryService.createWarehouse(
                request,
                actorId(jwt)
        );
    }

    @GetMapping("/warehouses/{warehouseId}/stock")
    public List<InventoryStockResponse> stock(
            @PathVariable UUID warehouseId,
            @AuthenticationPrincipal Jwt jwt
    ) {
        return inventoryService.listStock(
                warehouseId,
                actorId(jwt)
        );
    }

    @PostMapping("/movements")
    public InventoryStockResponse movement(
            @Valid @RequestBody InventoryMovementRequest request,
            @AuthenticationPrincipal Jwt jwt
    ) {
        return inventoryService.registerMovement(
                request,
                actorId(jwt)
        );
    }

    @GetMapping("/warehouses/{warehouseId}/movements")
    public List<InventoryMovementResponse> history(
            @PathVariable UUID warehouseId,
            @AuthenticationPrincipal Jwt jwt
    ) {
        return inventoryService.history(
                warehouseId,
                actorId(jwt)
        );
    }

    private UUID actorId(Jwt jwt) {
        return UUID.fromString(jwt.getSubject());
    }
}

File: apps/api/src/inventory/inventory.controller.ts
import {
  Body,
  Controller,
  Get,
  Param,
  ParseUUIDPipe,
  Post,
  Req,
  UseGuards,
} from "@nestjs/common";

import {
  inventoryMovementRequestSchema,
  inventoryTransferRequestSchema,
  warehouseRequestSchema,
  type InventoryMovementRequest,
  type InventoryMovementResponse,
  type InventoryStock,
  type InventoryTransferRequest,
  type InventoryTransferResponse,
  type WarehouseRequest,
  type WarehouseResponse,
} from "@velora/contracts";

import {
  BearerAuthGuard,
} from "../auth/bearer-auth.guard.js";

import type {
  AuthPrincipal,
} from "../auth/security.js";

import {
  RequireRoles,
} from "../common/authz/roles.decorator.js";

import {
  RolesGuard,
} from "../common/authz/roles.guard.js";

import type {
  AuthenticatedRequest,
} from "../common/http/authenticated-request.js";

import {
  ApiHttpError,
} from "../common/http/api-http.error.js";

import {
  ZodValidationPipe,
} from "../common/http/zod-validation.pipe.js";

import {
  InventoryService,
} from "./inventory.service.js";

@Controller("api/inventory")
@UseGuards(
  BearerAuthGuard,
  RolesGuard,
)
@RequireRoles(
  "ADMIN",
  "STORE_MANAGER",
)
export class InventoryController {
  constructor(
    private readonly inventory:
      InventoryService,
  ) {}

  @Get("warehouses")
  warehouses(
    @Req()
    request:
      AuthenticatedRequest,
  ): Promise<WarehouseResponse[]> {
    return this.inventory
      .listWarehouses(
        this.principal(
          request,
        ),
      );
  }

  @Post("warehouses")
  createWarehouse(
    @Req()
    request:
      AuthenticatedRequest,

    @Body(
      new ZodValidationPipe(
        warehouseRequestSchema,
      ),
    )
    body:
      WarehouseRequest,
  ): Promise<WarehouseResponse> {
    return this.inventory
      .createWarehouse(
        this.principal(
          request,
        ),
        body,
      );
  }

  @Get("warehouses/:warehouseId/stock")
  stock(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "warehouseId",
      new ParseUUIDPipe(),
    )
    warehouseId: string,
  ): Promise<InventoryStock[]> {
    return this.inventory
      .listStock(
        this.principal(
          request,
        ),
        warehouseId,
      );
  }

  @Post("movements")
  movement(
    @Req()
    request:
      AuthenticatedRequest,

    @Body(
      new ZodValidationPipe(
        inventoryMovementRequestSchema,
      ),
    )
    body:
      InventoryMovementRequest,
  ): Promise<InventoryStock> {
    return this.inventory
      .registerMovement(
        this.principal(
          request,
        ),
        body,
      );
  }

  @Post("transfers")
  transfer(
    @Req()
    request:
      AuthenticatedRequest,

    @Body(
      new ZodValidationPipe(
        inventoryTransferRequestSchema,
      ),
    )
    body:
      InventoryTransferRequest,
  ): Promise<InventoryTransferResponse> {
    return this.inventory
      .transfer(
        this.principal(
          request,
        ),
        body,
      );
  }

  @Get("warehouses/:warehouseId/movements")
  history(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "warehouseId",
      new ParseUUIDPipe(),
    )
    warehouseId: string,
  ): Promise<InventoryMovementResponse[]> {
    return this.inventory
      .history(
        this.principal(
          request,
        ),
        warehouseId,
      );
  }

  private principal(
    request:
      AuthenticatedRequest,
  ): AuthPrincipal {
    const principal =
      request.authPrincipal;

    if (!principal) {
      throw new ApiHttpError(
        401,
        "No autenticado.",
      );
    }

    return principal;
  }
}
